File: src/sdi.ts
import {SlippiGame} from './slippi'
import {Coord, CheckResult, Violation, isBoxController} from './index';

export enum SDIRegion {
    DZ = 0,
    NE = 1,
    SE = 2,
    SW = 3,
    NW = 4,
    N = 5,
    E = 6,
    S = 7,
    W = 8,
    TILT = 9,
}

const DIAGONALS = [SDIRegion.NE, SDIRegion.SE, SDIRegion.NW, SDIRegion.SW]
const CARDINALS = [SDIRegion.N, SDIRegion.E, SDIRegion.S, SDIRegion.W]

export function getSDIRegion(x: number, y: number): SDIRegion {
    if (Math.abs(x) <= 0.2875 && Math.abs(y) <= 0.2875) {
        return SDIRegion.DZ
    }

    let magnitude = Math.sqrt(x**2 + y**2)

    if (x >= 0.2875 && y >= 0.2875 && magnitude >= 0.7) {
        return SDIRegion.NE
    }
    if (x >= 0.2875 && y <= -0.2875 && magnitude >= 0.7) {
        return SDIRegion.SE
    } 
    if (x <= -0.2875 && y <= -0.2875 && magnitude >= 0.7) {
        return SDIRegion.SW
    } 
    if (x <= -0.2875 && y >= 0.2875 && magnitude >= 0.7) {
        return SDIRegion.NW
    }
    // Magnitude must be >=0.7 for these 4 if the condition is true. No need to check it
    if (y >= 0.7) {
        return SDIRegion.N
    } 
    if (x >= 0.7) {
        return SDIRegion.E
    } 
    if (y <= -0.7) {
        return SDIRegion.S
    } 
    if (x <= -0.7) {
        return SDIRegion.W
    }

    // Only place left
    return SDIRegion.TILT
}

// Directly adjacent
export function isRegionAdjacent(regionA: SDIRegion, regionB: SDIRegion): boolean {
    if (regionA == SDIRegion.N) {
        if (regionB == SDIRegion.NW || regionB == SDIRegion.NE) {
            return true
        }                 
    }
    if (regionA == SDIRegion.NE) {
        if (regionB == SDIRegion.N || regionB == SDIRegion.E) {
            return true
        }                 
    }
    if (regionA == SDIRegion.E) {
        if (regionB == SDIRegion.NE || regionB == SDIRegion.SE) {
            return true
        }                 
    }    
    if (regionA == SDIRegion.SE) {
        if (regionB == SDIRegion.E || regionB == SDIRegion.S) {
            return true
        }                 
    }
    if (regionA == SDIRegion.S) {
        if (regionB == SDIRegion.SE || regionB == SDIRegion.SW) {
            return true
        }                 
    }
    if (regionA == SDIRegion.SW) {
        if (regionB == SDIRegion.S || regionB == SDIRegion.W) {
            return true
        }                 
    }
    if (regionA == SDIRegion.W) {
        if (regionB == SDIRegion.SW || regionB == SDIRegion.NW) {
            return true
        }                 
    }
    if (regionA == SDIRegion.NW) {
        if (regionB == SDIRegion.W || regionB == SDIRegion.N) {
            return true
        }                 
    }
    return false
}


// Just for diagonals, ie: skipping the cardinals
export function isDiagonalAdjacent(regionA: SDIRegion, regionB: SDIRegion): boolean {
    const DIAGONALS = [SDIRegion.NE, SDIRegion.SE, SDIRegion.NW, SDIRegion.SW]
    if (!DIAGONALS.includes(regionA)) {
        return false
    }
    if (!DIAGONALS.includes(regionB)) {
        return false
    }

    if (regionA == SDIRegion.NE) {
        if (regionB == SDIRegion.NW || regionB == SDIRegion.SE) {
            return true
        }                 
    }
    if (regionA == SDIRegion.NW) {
        if (regionB == SDIRegion.NE || regionB == SDIRegion.SW) {
            return true
        }                 
    }
    if (regionA == SDIRegion.SW) {
        if (regionB == SDIRegion.SE || regionB == SDIRegion.NW) {
            return true
        }                 
    }
    if (regionA == SDIRegion.SE) {
        if (regionB == SDIRegion.NE || regionB == SDIRegion.SW) {
            return true
        }                 
    }

    return false
}
  
// Rapidly tapping the same direction and returning to neutral faster than once every 5.5 frames triggers 1 SDI and ignores subsequent attempts.
// Returns: array of violations
export function failsSDIRuleOne(coords: Coord[]): Violation[] {
    // Pull out the region of every input
    let violations: Violation[] = []
    let regions: SDIRegion[] = []
    for (let coord of coords) {
        regions.push(getSDIRegion(coord.x, coord.y))
    }

    for (let [i, region] of regions.entries()) {
        // Look ahead 10 frames to see if we hit two SDIs from neutral
        if (region === SDIRegion.DZ) {
            let lastRegion: SDIRegion = SDIRegion.DZ
            let firstSDIRegion: SDIRegion = null
            let lastSDIFrame: number = -1000
            let consecutiveTiltFrames = 0
            let hasTouchedDZ: boolean = true // You have to touch the deadzone in order for an SDI to count
            for (let j = 1; j <= 9 && (i+j) < regions.length; j++) {
                if (regions[i+j] === SDIRegion.DZ) {
                    hasTouchedDZ = true
                }
                if (regions[i+j] === SDIRegion.TILT) {
                    consecutiveTiltFrames += 1
                } else {
                    // Get the first SDI region
                    if (regions[i+j] !== SDIRegion.DZ && firstSDIRegion === null) {
                        firstSDIRegion = regions[i+j]
                    }
                }
                // If we went from DZ to the first SDI region. 
                // And we also haven't spent more than 3 consecutive frames in the tilt zone
                if (hasTouchedDZ && [SDIRegion.DZ, SDIRegion.TILT].includes(lastRegion) && regions[i+j] === firstSDIRegion && consecutiveTiltFrames <= 3) {
                    if (i+j <= lastSDIFrame + 4){
                        // Two SDI frames were less than 5 frames away from each other!
                        violations.push(new Violation(i, "Failed SDI rule #1", coords.slice(i, i+10)))
                    }
                    lastSDIFrame = i+j
                    hasTouchedDZ = false // Reset the DZ counter
                }
                lastRegion = regions[i+j]

                if (regions[i+j] !== SDIRegion.TILT) {
                    consecutiveTiltFrames = 0
                }
            }
        }
    }

    return violations
}

// Rapidly tapping the same diagonal and returning to an adjacent cardinal faster than once every 5.5 frames triggers 1 SDI and ignores subsequent attempts.
// Returns: array of violations
export function failsSDIRuleTwo(coords: Coord[]): Violation[] {
    // Pull out the region of every input
    let violations: Violation[] = []
    let regions: SDIRegion[] = []
    for (let coord of coords) {
        regions.push(getSDIRegion(coord.x, coord.y))
    }
    
    for (let i = 0; i < regions.length; i++) {
        const startingRegion = regions[i];

        // Start from a cardinal
        if (!CARDINALS.includes(startingRegion)) {
            continue
        }

        // Now look 4 frames ahead. 
        // Do we alternate between here and an adjacent diagonal twice?
        let sdiCount = 0
        let adjacentCardinalRegion = -1
        for (let j = 1; j <= 4 && (i+j) < regions.length; j++) {
            // Ignore if we haven't moved regions
            if (regions[i+j] === regions[i+j-1]) {
                continue
            }

            // Have we hit the diagonal SDI? (an adjacent diagonal)
            if (isRegionAdjacent(startingRegion, regions[i+j]) && DIAGONALS.includes(regions[i+j])) {
                if (adjacentCardinalRegion === -1 || adjacentCardinalRegion === regions[i+j]) {
                    adjacentCardinalRegion = regions[i+j]
                    sdiCount++
                }
            }
        }
        if (sdiCount >= 2) {
            violations.push(new Violation(i, "Failed SDI rule #2", coords.slice(i, i+5)))
        }
    }
    return violations
}

// Alternating between adjacent cardinals
// Returns: array of violations
export function failsSDIRuleThree(coords: Coord[]): Violation[] {
    // Pull out the region of every input
    let violations: Violation[] = []
    let regions: SDIRegion[] = []
    for (let coord of coords) {
        regions.push(getSDIRegion(coord.x, coord.y))
    }
    
     for (let i = 0; i < regions.length; i++) {
        const currentRegion = regions[i];

        if (DIAGONALS.includes(currentRegion)) {
            // Look forward 4 frames to see if it goes to an adjacent diagonal and back 
            var hitAdjacent = false
            for (let j = i + 1; j <= i + 4 && j < regions.length; j++) {
                // Hit the adjacent
                if (isDiagonalAdjacent(regions[j], currentRegion)) {
                    hitAdjacent = true
                }
                // Then returned back
                if (hitAdjacent && (regions[j] === currentRegion)) {
                    violations.push(new Violation(i, "Failed SDI rule #3", coords.slice(i, i+5)))
                }
            }
        }
    }
  
    return violations;
  }

export function hasIllegalSDI(game: SlippiGame, playerIndex: number, coords: Coord[]): CheckResult {
    // If we're on analog, then it always passes
    if (!isBoxController(coords)) {
        return new CheckResult(false)
    }

    let infractionFrames = failsSDIRuleOne(coords)
    if (infractionFrames.length > 0) {
        return new CheckResult(true, infractionFrames)
    } 
     
    infractionFrames = failsSDIRuleTwo(coords)
    if (infractionFrames.length > 0) {
        return new CheckResult(true, infractionFrames)
    } 

    infractionFrames = failsSDIRuleThree(coords)
    if (infractionFrames.length > 0) {
        return new CheckResult(true, infractionFrames)
    } 

    return new CheckResult(false)
}

File: libenforcer-wasm/src/checks/sdi.rs
use crate::types::{CheckResult, Coord, Violation};
use crate::utils::is_box_controller;
use std::collections::HashSet;

/// SDI regions for directional input classification
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum SDIRegion {
    DZ = 0,    // Deadzone
    NE = 1,    // Northeast diagonal
    SE = 2,    // Southeast diagonal
    SW = 3,    // Southwest diagonal
    NW = 4,    // Northwest diagonal
    N = 5,     // North cardinal
    E = 6,     // East cardinal
    S = 7,     // South cardinal
    W = 8,     // West cardinal
    TILT = 9,  // Tilt zone (between deadzone and full directions)
}

const DIAGONALS: [SDIRegion; 4] = [
    SDIRegion::NE,
    SDIRegion::SE,
    SDIRegion::SW,
    SDIRegion::NW,
];

const CARDINALS: [SDIRegion; 4] = [
    SDIRegion::N,
    SDIRegion::E,
    SDIRegion::S,
    SDIRegion::W,
];

/// Classify a coordinate into an SDI region
pub fn get_sdi_region(x: f64, y: f64) -> SDIRegion {
    // Deadzone
    if x.abs() <= 0.2875 && y.abs() <= 0.2875 {
        return SDIRegion::DZ;
    }

    let magnitude = (x.powi(2) + y.powi(2)).sqrt();

    // Diagonals (need both X and Y beyond threshold AND sufficient magnitude)
    if x >= 0.2875 && y >= 0.2875 && magnitude >= 0.7 {
        return SDIRegion::NE;
    }
    if x >= 0.2875 && y <= -0.2875 && magnitude >= 0.7 {
        return SDIRegion::SE;
    }
    if x <= -0.2875 && y <= -0.2875 && magnitude >= 0.7 {
        return SDIRegion::SW;
    }
    if x <= -0.2875 && y >= 0.2875 && magnitude >= 0.7 {
        return SDIRegion::NW;
    }

    // Cardinals (magnitude must be >= 0.7 if we reach here)
    if y >= 0.7 {
        return SDIRegion::N;
    }
    if x >= 0.7 {
        return SDIRegion::E;
    }
    if y <= -0.7 {
        return SDIRegion::S;
    }
    if x <= -0.7 {
        return SDIRegion::W;
    }

    // Everything else is tilt zone
    SDIRegion::TILT
}

/// Check if two regions are directly adjacent
pub fn is_region_adjacent(region_a: SDIRegion, region_b: SDIRegion) -> bool {
    use SDIRegion::*;
    match region_a {
        N => matches!(region_b, NW | NE),
        NE => matches!(region_b, N | E),
        E => matches!(region_b, NE | SE),
        SE => matches!(region_b, E | S),
        S => matches!(region_b, SE | SW),
        SW => matches!(region_b, S | W),
        W => matches!(region_b, SW | NW),
        NW => matches!(region_b, W | N),
        _ => false,
    }
}

/// Check if two diagonal regions are adjacent (skipping cardinals)
pub fn is_diagonal_adjacent(region_a: SDIRegion, region_b: SDIRegion) -> bool {
    use SDIRegion::*;

    if !DIAGONALS.contains(&region_a) || !DIAGONALS.contains(&region_b) {
        return false;
    }

    match region_a {
        NE => matches!(region_b, NW | SE),
        NW => matches!(region_b, NE | SW),
        SW => matches!(region_b, SE | NW),
        SE => matches!(region_b, NE | SW),
        _ => false,
    }
}

/// Count unique coordinates in a slice
fn count_unique_coordinates(coords: &[Coord]) -> usize {
    let mut unique = HashSet::new();
    for coord in coords {
        unique.insert((coord.x.to_bits(), coord.y.to_bits()));
    }
    unique.len()
}

/// SDI Rule #1: Rapidly tapping the same direction and returning to neutral
/// faster than once every 5.5 frames triggers 1 SDI and ignores subsequent attempts
pub fn fails_sdi_rule_one(coords: &[Coord]) -> Vec<Violation> {
    let mut violations = Vec::new();

    // Convert all coords to regions
    let regions: Vec<SDIRegion> = coords.iter().map(|c| get_sdi_region(c.x, c.y)).collect();

    for (i, &region) in regions.iter().enumerate() {
        // Look ahead 10 frames when starting from deadzone
        if region != SDIRegion::DZ {
            continue;
        }

        let mut last_region = SDIRegion::DZ;
        let mut first_sdi_region: Option<SDIRegion> = None;
        let mut last_sdi_frame: i32 = -1000;
        let mut consecutive_tilt_frames = 0;
        let mut has_touched_dz = true; // Must touch deadzone for SDI to count

        for j in 1..=9 {
            if i + j >= regions.len() {
                break;
            }

            let current_region = regions[i + j];

            if current_region == SDIRegion::DZ {
                has_touched_dz = true;
            }

            if current_region == SDIRegion::TILT {
                consecutive_tilt_frames += 1;
            } else {
                // Get the first SDI region
                if current_region != SDIRegion::DZ && first_sdi_region.is_none() {
                    first_sdi_region = Some(current_region);
                }
            }

            // If we went from DZ/TILT to the first SDI region
            // And haven't spent more than 3 consecutive frames in tilt zone
            if has_touched_dz
                && (last_region == SDIRegion::DZ || last_region == SDIRegion::TILT)
                && Some(current_region) == first_sdi_region
                && consecutive_tilt_frames <= 3
            {
                let current_frame = (i + j) as i32;

                if current_frame <= last_sdi_frame + 4 {
                    // This is a hack to be lenient as long as there's travel time
                    if count_unique_coordinates(&coords[i..i + j]) <= 2 {
                        // Two SDI frames were less than 5 frames away from each other!
                        let evidence = coords[i..(i + 10).min(coords.len())].to_vec();
                        violations.push(Violation::with_evidence(
                            i as f64,
                            "Failed SDI rule #1".to_string(),
                            evidence,
                        ));
                    }
                }

                last_sdi_frame = current_frame;
                has_touched_dz = false; // Reset the DZ counter
            }

            last_region = current_region;

            if current_region != SDIRegion::TILT {
                consecutive_tilt_frames = 0;
            }
        }
    }

    violations
}

/// SDI Rule #2: Rapidly tapping the same diagonal and returning to an adjacent cardinal
/// faster than once every 5.5 frames
pub fn fails_sdi_rule_two(coords: &[Coord]) -> Vec<Violation> {
    let mut violations = Vec::new();

    let regions: Vec<SDIRegion> = coords.iter().map(|c| get_sdi_region(c.x, c.y)).collect();

    for i in 0..regions.len() {
        let starting_region = regions[i];

        // Start from a cardinal
        if !CARDINALS.contains(&starting_region) {
            continue;
        }

        // Look 4 frames ahead
        // Do we alternate between here and an adjacent diagonal twice?
        let mut sdi_count = 0;
        let mut adjacent_diagonal_region: Option<SDIRegion> = None;

        for j in 1..=4 {
            if i + j >= regions.len() {
                break;
            }

            // Ignore if we haven't moved regions
            if regions[i + j] == regions[i + j - 1] {
                continue;
            }

            // Have we hit the diagonal SDI? (an adjacent diagonal)
            if is_region_adjacent(starting_region, regions[i + j])
                && DIAGONALS.contains(&regions[i + j])
            {
                if adjacent_diagonal_region.is_none()
                    || adjacent_diagonal_region == Some(regions[i + j])
                {
                    adjacent_diagonal_region = Some(regions[i + j]);
                    sdi_count += 1;
                }
            }
        }

        if sdi_count >= 2 {
            let evidence = coords[i..(i + 5).min(coords.len())].to_vec();
            violations.push(Violation::with_evidence(
                i as f64,
                "Failed SDI rule #2".to_string(),
                evidence,
            ));
        }
    }

    violations
}

/// SDI Rule #3: Alternating between adjacent diagonals
pub fn fails_sdi_rule_three(coords: &[Coord]) -> Vec<Violation> {
    let mut violations = Vec::new();

    let regions: Vec<SDIRegion> = coords.iter().map(|c| get_sdi_region(c.x, c.y)).collect();

    for i in 0..regions.len() {
        let current_region = regions[i];

        if !DIAGONALS.contains(&current_region) {
            continue;
        }

        // Look forward 4 frames to see if it goes to an adjacent diagonal and back
        let mut hit_adjacent = false;

        for j in (i + 1)..=(i + 4).min(regions.len() - 1) {
            // Hit the adjacent diagonal
            if is_diagonal_adjacent(regions[j], current_region) {
                hit_adjacent = true;
            }

            // Then returned back
            if hit_adjacent && regions[j] == current_region {
                let evidence = coords[i..(i + 5).min(coords.len())].to_vec();
                violations.push(Violation::with_evidence(
                    i as f64,
                    "Failed SDI rule #3".to_string(),
                    evidence,
                ));
                break;
            }
        }
    }

    violations
}

/// Check for illegal SDI patterns
/// Only applies to box controllers
pub fn check(coords: &[Coord]) -> CheckResult {
    // Only applies to box controllers
    if !is_box_controller(coords) {
        return CheckResult::pass();
    }

    let mut all_violations = Vec::new();

    all_violations.extend(fails_sdi_rule_one(coords));
    all_violations.extend(fails_sdi_rule_two(coords));
    all_violations.extend(fails_sdi_rule_three(coords));

    if all_violations.is_empty() {
        CheckResult::pass()
    } else {
        CheckResult::fail(all_violations)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sdi_region_classification() {
        assert_eq!(get_sdi_region(0.0, 0.0), SDIRegion::DZ);
        assert_eq!(get_sdi_region(0.9, 0.9), SDIRegion::NE);
        assert_eq!(get_sdi_region(0.9, 0.0), SDIRegion::E);
        assert_eq!(get_sdi_region(0.0, 0.9), SDIRegion::N);
        assert_eq!(get_sdi_region(-0.9, -0.9), SDIRegion::SW);
    }

    #[test]
    fn test_region_adjacency() {
        assert!(is_region_adjacent(SDIRegion::N, SDIRegion::NE));
        assert!(is_region_adjacent(SDIRegion::N, SDIRegion::NW));
        assert!(!is_region_adjacent(SDIRegion::N, SDIRegion::S));
    }

    #[test]
    fn test_diagonal_adjacency() {
        assert!(is_diagonal_adjacent(SDIRegion::NE, SDIRegion::NW));
        assert!(is_diagonal_adjacent(SDIRegion::NE, SDIRegion::SE));
        assert!(!is_diagonal_adjacent(SDIRegion::NE, SDIRegion::SW));
    }
}
